### ISSUE TITLE
Rule proposal: no-async-in-render

### ISSUE THREAD

**Issue by aequasi** [yannickcr/eslint-plugin-react#2550]
I'd like a rule that warns if there is an async method in the body of a render method, or outside of a hook:

```typescript jsx
class Foo extends React.Component {
    async bar() {
        await fetch('/bar');
    }

    render() {
        this.bar(); // Warn about this

        return <h1>Foo</h1>;
    }
}
```

```typescript jsx
export default function Foo() {
    async function bar() {
        await fetch('/bar');
    }

    bar(); // Warn about this

    return <h1>Foo</h1>;
}
```

**Comment by ljharb** [comment]
1. Why would you want to warn on that? It all seems fine to me.
1. What about a function that returns a Promise, or an `async function` in another file or a superclass (that wouldn't be statically detectable)?

**Comment by aequasi** [comment]
Because the async function would happen on every render. This is something that should very likely be in a `componentDidUpdate` or in a `useEffect`.

I'm not sure what is or isnt capable with eslint, and what it can detect. If its not possible, then we can close this out.

**Comment by ljharb** [comment]
It's possible to detect the specific case of running an `async function`, declared in the same file or class, inside a render path - but it would have lots of false negatives, because an asynchronous function in general wouldn't be warned on, anything declared in another file wouldn't be detectable, and it wouldn't be possible to exempt async calls in a custom hook.

There's lots of things that shouldn't happen on every render, sync or async, and I'm not sure it's really tenable to handle those outside of manual code review.

**Comment by aequasi** [comment]
Alright
